Proofs/SHA512/SHA512Loop.lean

```diff
@@ -14,110 +16,288 @@ is satisfied.
+  r (.SFP 7#5) sf =
+    extractLsb' 64 128
+      (DPSFP.vrev128_64_8 (read_mem_bytes 16 (InputBase + 0x50#64) si) ++
+        DPSFP.vrev128_64_8 (read_mem_bytes 16 (InputBase + 0x40#64) si)) ∧
+  r (.SFP 16#5) sf =
+    DPSFP.sha512su1 (DPSFP.vrev128_64_8 (read_mem_bytes 16 (InputBase + 0x70#64) si))
+      (extractLsb' 64 128
+        (DPSFP.vrev128_64_8 (read_mem_bytes 16 (InputBase + 0x50#64) si) ++
+          DPSFP.vrev128_64_8 (read_mem_bytes 16 (InputBase + 0x40#64) si)))
+      (DPSFP.sha512su0 (DPSFP.vrev128_64_8 (read_mem_bytes 16 (InputBase + 0x10#64) si))
+        (DPSFP.vrev128_64_8 (read_mem_bytes 16 InputBase si))) ∧
+  r (.SFP 24#5) sf =
+    extractLsb' 64 128
+      (DPSFP.binary_vector_op_aux 0 2 64 BitVec.add (r (StateField.SFP 0x18#5) si)
+          (DPSFP.vrev128_64_8 (read_mem_bytes 16 InputBase si)) 0x0#128 ++
+        DPSFP.binary_vector_op_aux 0 2 64 BitVec.add (r (StateField.SFP 0x18#5) si)
+          (DPSFP.vrev128_64_8 (read_mem_bytes 16 InputBase si)) 0x0#128) ∧
+  r (.SFP 25#5) sf =
+    read_mem_bytes 16 (KtblAddr + 0x10#64) si
 
+/--
+info: #[RegType.SFP 0x03#5, RegType.SFP 0x04#5, RegType.SFP 0x05#5, RegType.SFP 0x06#5, RegType.SFP 0x07#5,
+  RegType.SFP 0x10#5, RegType.SFP 0x18#5, RegType.SFP 0x19#5]
+-/
+#guard_msgs in
+#eval ((Cfg.create' (0x126500#64 + 8*4) (0x126500#64 + 8*4 + 12*4) SHA512.program).toOption.get!).maybe_modified_regs
+
+theorem sha512_block_armv8_loop_2 (sprev si sf : ArmState)
+  (h_N : N = 1#64)
+  (h_si_prelude : loop_post_1 (0x126500#64 + 8*4) N SP CtxBase InputBase sprev si)
+  (h_steps : nsteps = 12)
+  (h_run : sf = run nsteps si) :
+  loop_post_2 (0x126500#64 + 8*4 + nsteps*4) N SP CtxBase InputBase
+              si sf := by
+  -- Prelude
+  subst h_N h_steps
+  simp at h_si_prelude ⊢
+  obtain ⟨h_si_num_blocks, h_si_program, h_si_pc, h_si_err,
+          h_si_sp_aligned, h_si_input_base, h_si_ctx_base,
+          h_si_sp, h_si_ktbl, h_si_input,
+          h_si_separate, h_si_ktbl_base,
+          _h_si_q0, _h_si_q1, _h_si_q2, _h_si_q3,
+          h_si_16, h_si_17, h_si_18, h_si_19,
+          h_si_20, h_si_21, h_si_22, h_si_23, h_si_zf⟩ := h_si_prelude
+  simp only [input_addr, ctx_addr] at *
+  simp only [BitVec.reduceToNat, Nat.reduceMul] at h_si_input
+  simp only [loop_post_2]
+  -- Symbolic Simulation
+  sym_n 12
   -- Epilogue
-  -- cse (config := { processHyps := .allHyps })
-  -- simp (config := {ground := true}) only
-        -- [fst_AddWithCarry_eq_sub_neg,
-        --  ConditionHolds,
-        --  state_simp_rules,
-        --  bitvec_rules, minimal_theory]
-  -- sym_aggregate
-  -- assumption
-  -- done
-  sorry
+  simp only [←Memory.mem_eq_iff_read_mem_bytes_eq] at *
+  simp only [memory_rules] at *
+  simp at h_si_separate
+  simp only [h_si_ktbl, h_si_separate, minimal_theory]
+  have h_si_input_1 : (Memory.read_bytes 16 InputBase sprev.mem) =
+                      (Memory.read_bytes 16 InputBase si.mem) := by
+    -- simp_mem (config := {useOmegaToClose := true})
```

Comment:
@bollu What am I doing wrong that `simp_mem` isn't helping here?

Reply:
From a conversation with @bollu: looks like `omega` is generating a gigantic proof, which causes `simp_mem` to fail here. We'll table this for now.